File: Summary.py
import win32com.client as win32
import pandas as pd

# Define the date range for filter
start_date = '01-01-2023'
end_date = '30-12-2023'

# Open the Excel workbook
excel = win32.gencache.EnsureDispatch('Excel.Application')
workbook = excel.Workbooks.Open(r'https://my.sharepoint.com/yourfile.xlsm')

# Filter the dataframe by the desired date range
workbook = workbook.loc[(workbook['date_column'] >= start_date) & (workbook['date_column'] <= end_date)]

# Check if Summary sheet already exist, if not create one
if "Summary" not in [sheet.Name for sheet in workbook.Sheets]:
    summary_sheet = workbook.Sheets.Add(After=workbook.Sheets(workbook.Sheets.Count))
    summary_sheet.Name = "Summary"
else:
    summary_sheet = workbook.Sheets("Summary")
    summary_sheet.UsedRange.Clear()

# Loop through all sheets in the workbook (excluding "Summary" sheet)
for sheet in workbook.Sheets:
    if sheet.Name != "Summary":
        # Copy headers
        sheet.Range(sheet.Rows(1), sheet.Rows(1)).Copy()
        summary_sheet.Range(summary_sheet.Rows(summary_sheet.Rows.Count), summary_sheet.Rows(summary_sheet.Rows.Count)).Offset(1, 0).PasteSpecial()
        # Get last row of data for today's date
        last_row = sheet.Cells(sheet.Rows.Count, 1).End(win32.constants.xlUp).Row
        # Loop through each row in the current sheet
        for row in range(2, last_row + 1):
            # Check if the date in the first column matches today's date
            if sheet.Cells(row, 1).Value == date_filter:
                # Copy the current row and paste it into the "Summary" sheet
                sheet.Range(sheet.Rows(row), sheet.Rows(row)).Copy()
                summary_sheet.Range(summary_sheet.Rows(summary_sheet.Rows.Count), summary_sheet.Rows(summary_sheet.Rows.Count)).Offset(1, 0).PasteSpecial()
                # Set the number format for the date column
                summary_sheet.Cells(summary_sheet.Rows.Count, 1).End(win32.constants.xlUp).NumberFormat = "dd/mm/yyyy"

# Autofit the columns
summary_sheet.Columns.AutoFit()

# Save the workbook
workbook.Save()

# Close Excel
excel.Application.Quit()

File: Summary.vb
Sub CreateSummary()
    ' Get today's date
    Dim today As String
    today = Format(Date, "dd/MM/yyyy")
    today = Format(Date, "dd/MM/yyyy", vbUseSystemDayOfWeek, vbUseSystem)
    
    ' Check if Summary sheet already exist, if not create one
    Dim summarySheet As Worksheet
    On Error Resume Next
    Set summarySheet = ThisWorkbook.Sheets("Summary")
    On Error GoTo 0
    If summarySheet Is Nothing Then
        Set summarySheet = ThisWorkbook.Sheets.Add(After:=ThisWorkbook.Sheets(ThisWorkbook.Sheets.Count))
        summarySheet.Name = "Summary"
    Else
        ' Clear existing data in Summary sheet
        summarySheet.UsedRange.Clear
    End If

    ' Loop through all sheets in the workbook (excluding "Summary" sheet)
    Dim ws As Worksheet
    Dim lastRow As Long
    Dim r As Long
    Dim sheetCounter As Integer
        sheetCounter = 0
    For Each ws In ThisWorkbook.Sheets
        If ws.Name <> "Summary" And sheetCounter < 5 Then
            sheetCounter = sheetCounter + 1
            'Add the sheet name to the summary sheet
            r = summarySheet.Cells(summarySheet.Rows.Count, 1).End(xlUp).Row + 1
            summarySheet.Cells(r, 1).Value = ws.Name
            summarySheet.Cells(r, 1).Font.Bold = True
            summarySheet.Cells(r, 1).Font.Size = 14
            summarySheet.Cells(r, 1).Interior.Color = RGB(255, 153, 0)

            ' Copy headers
            ws.Rows(1).Copy
            summarySheet.Cells(summarySheet.Rows.Count, "A").End(xlUp).Offset(1, 0).PasteSpecial xlPasteValues
            ' Set headers to bold
            summarySheet.Cells(summarySheet.Rows.Count, "A").End(xlUp).Resize(1, ws.UsedRange.Columns.Count).Font.Bold = True
            ' Get last row of data for today's date
            lastRow = ws.Cells(ws.Rows.Count, 1).End(xlUp).Row
            ' Loop through each row in the current sheet
            For i = 2 To lastRow
                ' Check if the date in the first column matches today's date
                If ws.Cells(i, 1).Value = Date Then
                    ' Copy the current row and paste it into the "Summary" sheet
                    ws.Rows(i).Copy
                    summarySheet.Cells(summarySheet.Rows.Count, "A").End(xlUp).Offset(1, 0).PasteSpecial xlPasteValues
                    ' Set the number format for the date column
                    summarySheet.Cells(summarySheet.Rows.Count, 1).End(xlUp).NumberFormat = "dd/mm/yyyy"
                End If
            Next i
        End If
    Next ws
    ' Autofit the columns
    summarySheet.Columns.AutoFit
End Sub
